Adds sql creation scripts for postgres

vishalkuo opened the pull request:
Referencing https://github.com/google/trillian/issues/1298, this PR adds the SQL creation script for a postgres data store and a stripped down reset script. I've done my best to translate the MySQL <> Postgres types; however they may not be a perfect mapping and it's possible that there'll need to be further changes made but I figured it's a good starting point for now. 

Note that I switched to camel casing since Postgres is case insensitive and would normalize everything to being lowercased which is hard to read without underscores.

gdbelvin commented on storage/postgres/storage.sql:
```diff
@@ -0,0 +1,138 @@
+-- are allowed they will all reference this row.
+CREATE TABLE IF NOT EXISTS leaf_data(
+  tree_id               BIGINT NOT NULL,
+  -- This is a personality specific has of some subset of the leaf data.
```
s/has/hash/

gdbelvin commented on storage/postgres/storage.sql:
```diff
@@ -0,0 +1,138 @@
+  -- This is a personality specific has of some subset of the leaf data.
+  -- It's only purpose is to allow Trillian to identify duplicate entries in
+  -- the context of the personality.
+  leaf_identity_hash     BIT VARYING(255) NOT NULL,
```
should this be `BYTEA` or `BYTEA(32)`?

AlCutter replied:
In other storage schemas, we've deliberately kept the field size overly large in order to accommodate > 256bit hashes for future expansion.

gdbelvin commented on storage/postgres/storage.sql:
```diff
@@ -0,0 +1,138 @@
+  -- This is a personality specific has of some subset of the leaf data.
+  -- It's only purpose is to allow Trillian to identify duplicate entries in
+  -- the context of the personality.
+  leaf_identity_hash        BIT VARYING(255) NOT NULL,
```
Same `BYTEA` vs `BIT` question

AlCutter commented on storage/postgres/storage.sql:
```diff
@@ -0,0 +1,138 @@
+
+CREATE TABLE IF NOT EXISTS subtree(
+  tree_id               BIGINT NOT NULL,
+  subtree_id            BIT VARYING(255) NOT NULL,
```
Perhaps BYTEA here too (and consider size)

AlCutter commented on storage/postgres/storage.sql:
```diff
@@ -0,0 +1,138 @@
+  tree_id                BIGINT NOT NULL,
+  tree_head_timestamp    BIGINT,
+  tree_size              BIGINT,
+  root_hash              BIT VARYING(255) NOT NULL,
```
Perhaps BYTEA here too (and consider size)

AlCutter commented on storage/postgres/storage.sql:
```diff
@@ -0,0 +1,138 @@
+  tree_head_timestamp    BIGINT,
+  tree_size              BIGINT,
+  root_hash              BIT VARYING(255) NOT NULL,
+  root_signature         BIT VARYING(1024) NOT NULL,
```
Perhaps BYTEA here too (and consider size)

AlCutter commented on storage/postgres/storage.sql:
```diff
@@ -0,0 +1,138 @@
+  leaf_value            BYTEA NOT NULL,
+  -- This is extra data that the application can associate with the leaf should it wish to.
+  -- This data is not included in signing and hashing.
+  extra_dta            BYTEA,
```
Consider `extra_data` to mirror naming elsewhere?

AlCutter commented on storage/postgres/storage.sql:
```diff
@@ -0,0 +1,138 @@
+  leaf_identity_hash        BIT VARYING(255) NOT NULL,
+  -- This is a MerkleLeafHash as defined by the treehasher that the log uses. For example for
+  -- CT this hash will include the leaf prefix byte as well as the leaf data.
+  merkle_leaf_hash          BIT VARYING(255) NOT NULL,
```
Perhaps BYTEA here too (and consider size)

AlCutter commented on storage/postgres/storage.sql:
```diff
@@ -0,0 +1,138 @@
+  leaf_identity_hash    BIT VARYING(255) NOT NULL,
+  -- This is a MerkleLeafHash as defined by the treehasher that the log uses. For example for
+  -- CT this hash will include the leaf prefix byte as well as the leaf data.
+  merkle_leaf_hash      BIT VARYING(255) NOT NULL,
```
Perhaps BYTEA here too (and consider size)

AlCutter commented on storage/postgres/storage.sql:
```diff
@@ -0,0 +1,138 @@
+  -- This is a SHA256 hash of the TreeID, LeafIdentityHash and QueueTimestampNanos. It is used
+  -- for batched deletes from the table when trillian_log_server and trillian_log_signer are
+  -- built with the batched_queue tag.
+  queue_id              BIT VARYING(32) DEFAULT NULL UNIQUE,
```
Perhaps BYTEA here too (and consider size)

vishalkuo commented:
@gdbelvin I've switched from `BIT VARYING` to `BYTEA`; however, it's worth noting that BYTEA doesn't allow one to specify length (although it has a limit of [1GB](https://stackoverflow.com/questions/26635547/postgresql-difference-between-bytea-and-bit-varying-types)). 

Regarding integration tests; are you referring to tests similar to those under the `integration` directory? If that's the case; it seems like I'll need to finish the rest of the pg implementation before being able to provide those. My plan was to use these types as a starting point and build `log_storage.go` and `admin_storage.go`, modifying these sql types as need be, until they reach a satisfactory state. Let me know if this isn't how you intended for this process to go and I can close this and come back with a full PR once I've got a locally working version.

gdbelvin commented:
Smaller PRs are easier to review.  Let's proceed with this and we can modify as needed when we get some unit tests working.

gdbelvin commented:
Thanks for pointing out the lack of a size limit on `BYTEA`.  
`BYTEA` is probably still more appropriate for hash outputs than `BIT` which is designed for bit masks.

AlCutter commented:
Thanks for the updates, I think it's probably ok to have `BYTEA` with no limit, the on-disk size requirements are essentially the same anyway - `len(DATA) + (1 or 4) bytes`, and the personality layer can enforce sensible lower size limits for leaves.

AlCutter commented on storage/postgres/storage.sql:
```diff
@@ -0,0 +1,138 @@
+  root_hash              BYTEA NOT NULL,
+  root_signature         BYTEA NOT NULL,
+  tree_revision          BIGINT,
+  PRIMARY KEY(tree_id, tree_head_timestamp),
```
I _think_ it would be better to have the PK be `tree_id, tree_revision`, you could then drop the `index` below (I don't think anything needs this key to contain the timestamp - the cloudspanner implementation uses `/treeID/tree_revision` as it's PK on the equivalent table there BTW).

Separately, and I must admit I don't know much about PGSQL, but other storage impls define the PK here as `tree_id, tree_rev DESC`, because by far the most common operation on this table is a scan which  wants to fetch the newest entry.  It looks like PG doesn't support the `DESC` modifier on the PK fields, so I don't know what implications this will have for performance when you come to write the code itself.

Might be worth popping a comment or a `TODO` here as a note to yourself to look at this area again in detail once the code lands too.

vishalkuo commented:
@AlCutter you're right, it seems like postgres doesn't support scan ordering for primary keys. That being said, the workaround seems to be to create a unique index as well, seen [here](https://stackoverflow.com/questions/45597101/primary-key-with-asc-or-desc-ordering). I left the index below the create table statement and added a `desc` modifier on `tree_revision` to scan the most recent changes first. I also added a TODO to benchmark this change once we have more code in place to make sure this change is acceptable. Let me know if that works
